
<img src="Tp1.png">


### Importer les bibliothèques **numpy** et **pandas** et importer functools de reduce

In [9]:
import pandas as pd
import numpy as np

# PART I

Explorer la dataset `BL-Flickr-Images-Book.csv` 

**NB:** faites attentionsau chemin

## 1 Exploration du dataset

In [60]:
 
 df =pd.read_csv('C:/Users/hed/Downloads/DM-master/DM-master/TP1/Datasets/BL-Flickr-Images-Book.csv')
df.head(5)


Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [35]:
## 2 Jeter un coup d'oeil dans le dataset

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 15 columns):
Identifier                8287 non-null int64
Edition Statement         773 non-null object
Place of Publication      8287 non-null object
Date of Publication       8106 non-null object
Publisher                 4092 non-null object
Title                     8287 non-null object
Author                    6509 non-null object
Contributors              8287 non-null object
Corporate Author          0 non-null float64
Corporate Contributors    0 non-null float64
Former owner              1 non-null object
Engraver                  0 non-null float64
Issuance type             8287 non-null object
Flickr URL                8287 non-null object
Shelfmarks                8287 non-null object
dtypes: float64(3), int64(1), object(11)
memory usage: 971.2+ KB


## 3 Qu'est ce que vous remarquez?

##### La fonction info () est utilisée pour imprimer un résumé concis d'un DataFrame. Cette méthode imprime des informations sur un DataFrame, y compris les dtypes d'index et de colonne, les valeurs non nulles et l'utilisation de la mémoire. <br> =>Il faut le suprime les donées qu'ils sont null.




## 4 Dropping unnecessary columns

Nous allons restreindre l'étude à seulement 7 colomnes pour cela nous allons supprimer 7 colonnes que nous jugeons unitiles por le moment :
```Python
to_drop = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']```

In [55]:
to_drop = ['Edition Statement',
           'Corporate Author',
           'Corporate Contributors',
           'Former owner',
           'Engraver',
           'Contributors',
           'Issuance type',
           'Shelfmarks']
df.drop(columns=to_drop,inplace=True)
df.head()


 


Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

In [50]:
df=df.set_index("Identifier")

Edition Statement  \
Identifier                                                      
206                                                       NaN   
216                                                       NaN   
218                                                       NaN   
472                                                       NaN   
480                              A new edition, revised, etc.   
481                             Fourth edition, revised, etc.   
519                                                       NaN   
667                                                       NaN   
874                                                       NaN   
1143                                                      NaN   
1280                                                      NaN   
1808                                                      NaN   
1905                                                      NaN   
1929                                                      NaN   
2836                                                      NaN   
2854                                                      NaN   
2956                                                      NaN   
2957                                                      NaN   
3017        Nueva edicion, anotada ... y continuada ... po...   
3131                                                      NaN   
4598                                                      NaN   
4884                                                      NaN   
4976        [Another edition.] Abdollatiphi historiæ Ægypt...   
5382                                                      NaN   
5385            [Another edition.] Illustrated by John Leech.   
5389                                       [Another edition.]   
5432                                                      NaN   
6036                                                      NaN   
6821                                                      NaN   
7521                                                      NaN   
...                                                       ...   
4053464                                                   NaN   
4063671                                                   NaN   
4072044                                                   NaN   
4077038                                                   NaN   
4079258                                                   NaN   
4079262                                                   NaN   
4112297                                                   NaN   
4112525                                                   NaN   
4112839                                                   NaN   
4113012                                                   NaN   
4113816                                                   NaN   
4114334                                                   NaN   
4114390                                                   NaN   
4114889                                                   NaN   
4114986     [Another edition.] [With a prologue and epilog...   
4115138                                                   NaN   
4116063                                                   NaN   
4117526                                                   NaN   
4117583                                                   NaN   
4117749                                                   NaN   
4117751                                                   NaN   
4117752                                                   NaN   
4156359                                                   NaN   
4157746                                                   NaN   
4157862                                                   NaN   
4158088                                                   NaN   
4158128                                                   NaN   
4159563                                                   NaN   
4159587                                                   NaN   
4160339       

In [56]:
df=df.set_index("Identifier")


df.head()


Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

## 6 Donner un appérçu sur la colonne `Date of Publication` 


In [61]:
df['Date of Publication'].head(25)

0            1879 [1878]
1                   1868
2                   1869
3                   1851
4                   1857
5                   1875
6                   1872
7                    NaN
8                   1676
9                   1679
10                  1802
11                  1859
12                  1888
13           1839, 38-54
14                  1897
15                  1865
16               1860-63
17                  1873
18                  1866
19                  1899
20                  1814
21                  1820
22                  1800
23    1847, 48 [1846-48]
24               [1897?]
Name: Date of Publication, dtype: object

> Remarque : Observer la date de publication qui correspond à l'index 206.
> Ca nécessite un nettoyage et une correction : il faut éliminer certains symboles 

## 7 Cleaning columns using the `.apply` function

### Ecrire une fonction qui permet de nettoyer la colonne des dates 

In [92]:
unwanted_characters = ['[', ',', '-']
# Completer la fonction suivante
def clean_dates(item):
    dop= str(item.loc['Date of Publication'])
    
    if dop == 'nan' or dop[0] == '[':
         return 1
    
    for character in unwanted_characters:
        if character in dop:
            character_index=dop.find(character)
            dop=dop.lstrip(character_index)
            
            
    
    return dop


df['Date of Publication'] = df.apply(clean_dates, axis = 1)
df['Date of Publication'].head(25)

0     1879 
1      1868
2      1869
3      1851
4      1857
5      1875
6      1872
7       NaN
8      1676
9      1679
10     1802
11     1859
12     1888
13     1839
14     1897
15     1865
16     1860
17     1873
18     1866
19     1899
20     1814
21     1820
22     1800
23     1847
24      NaN
Name: Date of Publication, dtype: object

In [79]:
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0               1879        S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [62]:
#Voici une autre méthode
#exécuter pour voir le résultat
unwanted_characters = ['[', ',', '-']

def clean_dates(dop):
    dop = str(dop)
    if dop.startswith('[') or dop == 'nan':
        return 'NaN'
    for character in unwanted_characters:
        if character in dop:
            character_index = dop.find(character)
            dop = dop[:character_index]
    return dop

df['Date of Publication'] = df['Date of Publication'].apply(clean_dates)
df.head(10)

Identifier              Edition Statement  \
0         206                            NaN   
1         216                            NaN   
2         218                            NaN   
3         472                            NaN   
4         480   A new edition, revised, etc.   
5         481  Fourth edition, revised, etc.   
6         519                            NaN   
7         667                            NaN   
8         874                            NaN   
9        1143                            NaN   

                  Place of Publication Date of Publication  \
0                               London               1879    
1             London; Virtue & Yorston                1868   
2                               London                1869   
3                               London                1851   
4                               London                1857   
5                               London                1875   
6                               London                1872   
7  pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
8                              London]                1676   
9                               London                1679   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   
5      William Macintosh  [The World in which I live, and my place in it...   
6             The Author  Lagonells. By the author of Darmayne (F. E. A....   
7                    NaN  The Coming of Spring, and other poems. By J. A...   
8                    NaN  A Warning to the inhabitants of England, and L...   
9                    NaN  A Satyr against Vertue. (A poem: supposed to b...   

          Author                                  Contributors  \
0          A. A.                               FORBES, Walter.   
1      A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
2      A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
3      A., E. S.                   Appleyard, Ernest Silvanus.   
4      A., E. S.                           BROOME, John Henry.   
5      A., E. S.                           BROOME, John Henry.   
6      A., F. E.                       ASHLEY, Florence Emily.   
7  A., J.|A., J.                 ANDREWS, J. - Writer of Verse   
8         Remaʿ.                                  ADAMS, Mary.   
9         A., T.                                 OLDHAM, John.   

   Corporate Author  Corporate Contributors Former owner  Engraver  \
0               NaN                     NaN          NaN       NaN   
1               NaN                     NaN          NaN       NaN   
2               NaN                     NaN          NaN       NaN   
3               NaN                     NaN          NaN       NaN   
4               NaN                     NaN          NaN       NaN   
5               NaN                     NaN          NaN       NaN   
6               NaN                     NaN          NaN       NaN   
7               NaN                     NaN          NaN       NaN   
8               NaN                     NaN          NaN       NaN   
9               NaN                     NaN          NaN       NaN   

  Issuance type                                         Flickr URL  \
0   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
1   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
2   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
3   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
4   monographic  http://www.flickr.com/photos/britishlibrary/ta...   
5  

In [ ]:
## 8 Observer la collone `Author`
### Completer la fonction `clean_author_names`qui permet de corriger les valeurs de cet colonne

In [115]:
# compléter la fonction suivante ( à la place de pass)
from functools import reduce
def clean_author_names(author):
    
    author = str(author)
    
    if author == 'nan':
        return 'NaN'
    
    author = author.split(',')

    if len(author) == 1:
        name = filter(lambda x: x.isalpha(), author[0])
        return reduce(lambda x, y: x + y, name)
    
    last_name, first_name = author[0], author[1]

    first_name = first_name[:first_name.find('-')] if '-' in first_name else first_name
    
    if first_name.endswith(('.', '.|')):
        parts = first_name.split('.')
        
        if len(parts) > 1:
            first_occurence = first_name.find('.')
            final_occurence = first_name.find('.', first_occurence + 1)
            first_name= first_name[:final_occurence]
            
        else:
            first_name= first_name[:first_name.find('.')]
    
    last_name = last_name.capitalize()
    
    return f'{first_name} {last_name}'


df['Author'] = df['Author'].apply(clean_author_names)


# Résultat attendu 
<img src="Datasets\capture42.png">

In [116]:
df.head()

Identifier Edition Statement      Place of Publication Date of Publication  \
0         206               NaN                    London               1879    
1         216               NaN  London; Virtue & Yorston                1868   

          Publisher                                              Title Author  \
0  S. Tinsley & Co.                          Walter Forbes. [a Novel.]     AA   
1      Virtue & Co.  All For Greed. [a Novel. The Dedication Signed...    AAA   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   

                          Shelfmarks  
0  British Library HMNTS 12641.b.30.  
1  British Library HMNTS 12626.cc.2.

##  Observer la colone `title`
### 9 Completer la fonction `clean_title`qui permet de corriger les valeurs de cet colonne

# Résultat attendu 
 <img src="Datasets/capture43.png">

In [128]:
## complétrer la fonction suivante ( à la place de pass)

def clean_title(title):
    
    if title == 'nan':
        return 'NaN'
    
    if title[0] == '[':
        pass
        
    if 'by' in title:
        pass
    elif 'By' in title:
        pass
        
    if '[' in title:
        pass

    title = title[:-2]
        
    title = list(map(str.capitalize, title.split()))
    return ' '.join(title)
    
df['Title'] = df['Title'].apply(clean_title)
df.head()

IndexError: string index out of range

# PART II
## 1 Cleaning entire dataset
### On considère le fichier suivant 'university_towns.txt'

In [11]:
!head Datasets/university_towns.txt

Alabama[edit]
Auburn (Auburn University)[1]
Florence (University of North Alabama)
Jacksonville (Jacksonville State University)[2]
Livingston (University of West Alabama)[2]
Montevallo (University of Montevallo)[2]
Troy (Troy University)[2]
Tuscaloosa (University of Alabama, Stillman College, Shelton State)[3][4]
Tuskegee (Tuskegee University)[5]
Alaska[edit]


In [13]:
with open('Datasets/university_towns.txt', 'r') as text:
    textfile = text.read()
    print(textfile)

Alabama[edit]
Auburn (Auburn University)[1]
Florence (University of North Alabama)
Jacksonville (Jacksonville State University)[2]
Livingston (University of West Alabama)[2]
Montevallo (University of Montevallo)[2]
Troy (Troy University)[2]
Tuscaloosa (University of Alabama, Stillman College, Shelton State)[3][4]
Tuskegee (Tuskegee University)[5]
Alaska[edit]
Fairbanks (University of Alaska Fairbanks)[2]
Arizona[edit]
Flagstaff (Northern Arizona University)[6]
Tempe (Arizona State University)
Tucson (University of Arizona)
Arkansas[edit]
Arkadelphia (Henderson State University, Ouachita Baptist University)[2]
Conway (Central Baptist College, Hendrix College, University of Central Arkansas)[2]
Fayetteville (University of Arkansas)[7]
Jonesboro (Arkansas State University)[8]
Magnolia (Southern Arkansas University)[2]
Monticello (University of Arkansas at Monticello)[2]
Russellville (Arkansas Tech University)[2]
Searcy (Harding University)[5]
California[edit]
Angwin (Pacific Union College

## 2 Créer un pandas data frame à partir d'un fichier txt
On veut créer un DataFrame nommé `towns_df` à partir du fichier 'university_towns.txt'

Nous voyons que nous avons des noms d'états périodiques suivis des villes universitaires de cet état: StateA TownA1 TownA2 StateB TownB1 TownB2 .... Si nous regardons la façon dont les noms d'états sont écrits dans le fichier, nous verrons qu'ils ont tous la sous-chaîne «[edit]» en eux.

Nous pouvons tirer parti de ce modèle en créant une liste de tuples (état, ville) et en enveloppant cette liste dans un DataFrame:


In [126]:
university_towns = []
with open('Datasets/university_towns.txt') as file:
    for line in file:
        if '[edit]' in line:
             # se rappeler de ce `state` jusqu'à ce qu'on trouve le suivant
            state=line
        else:
             # Sinon, nous avons une ville; garder `state` 
              university_towns.append((state, line))

university_towns[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [15]:
towns_df = pd.DataFrame(university_towns,
                        columns=['State', 'RegionName'])
towns_df.head()

State                                         RegionName
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

## 3 Compléter la fonction `clean_up` pour nettoyer les colonnes de `towns_df`

In [16]:
## complétrer la fonction suivante ( à la place de pass)
def clean_up(item):
    if '(' in item:
        pass
        return item[:item.find('(') - 1]
    
    if '[' in item:
        return item[:item.find('[')]
    

towns_df =  towns_df.applymap(clean_up)
towns_df

State     RegionName
0      Alabama         Auburn
1      Alabama       Florence
2      Alabama   Jacksonville
3      Alabama     Livingston
4      Alabama     Montevallo
..         ...            ...
512  Wisconsin    River Falls
513  Wisconsin  Stevens Point
514  Wisconsin       Waukesha
515  Wisconsin     Whitewater
516    Wyoming        Laramie

[517 rows x 2 columns]

# PART III ( OPTIONNELLE) 
## 1 Renaming columns and skipping rows
### 1.1 Convertir le fichier `olympics.csv` en dataframe pandas `olympics_df`

In [71]:
 
olympics_df = pd.read_csv('C:/Users/hed/Downloads/DM-master/DM-master/TP1/Datasets/olympics.csv')

# Afficher les 5 premières lignes
olympics_df.head(5)

0         1     2     3     4      5         6     7     8  \
0                NaN  ? Summer  01 !  02 !  03 !  Total  ? Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  ? Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

### 1.2 Supprimer la première ligne 

In [73]:
olympics_df.drop(olympics_df.index[1]) 

0         1     2     3     4  \
0                                             NaN  ? Summer  01 !  02 !  03 !   
2                                   Algeria (ALG)        12     5     2     8   
3                                 Argentina (ARG)        23    18    24    28   
4                                   Armenia (ARM)         5     1     2     9   
5                         Australasia (ANZ) [ANZ]         2     3     4     5   
6                       Australia (AUS) [AUS] [Z]        25   139   152   177   
7                                   Austria (AUT)        26    18    33    35   
8                                Azerbaijan (AZE)         5     6     5    15   
9                                   Bahamas (BAH)        15     5     2     5   
10                                  Bahrain (BRN)         8     0     0     1   
11                           Barbados (BAR) [BAR]        11     0     0     1   
12                                  Belarus (BLR)         5    12    24    39   
13                                  Belgium (BEL)        25    37    52    53   
14                                  Bermuda (BER)        17     0     0     1   
15                        Bohemia (BOH) [BOH] [Z]         3     0     1     3   
16                                 Botswana (BOT)         9     0     1     0   
17                                   Brazil (BRA)        21    23    30    55   
18                British West Indies (BWI) [BWI]         1     0     0     2   
19                             Bulgaria (BUL) [H]        19    51    85    78   
20                                  Burundi (BDI)         5     1     0     0   
21                                 Cameroon (CMR)        13     3     1     1   
22                                   Canada (CAN)        25    59    99   121   
23                                Chile (CHI) [I]        22     2     7     4   
24                              China (CHN) [CHN]         9   201   146   126   
25                                 Colombia (COL)        18     2     6    11   
26                               Costa Rica (CRC)        14     1     1     2   
27                        Ivory Coast (CIV) [CIV]        12     0     1     0   
28                                  Croatia (CRO)         6     6     7    10   
29                                 Cuba (CUB) [Z]        19    72    67    70   
30                                   Cyprus (CYP)         9     0     1     0   
..                                            ...       ...   ...   ...   ...   
118                         Sri Lanka (SRI) [SRI]        16     0     2     0   
119                                   Sudan (SUD)        11     0     1     0   
120                            Suriname (SUR) [E]        11     1     0     1   
121                              Sweden (SWE) [Z]        26   143   164   176   
122                             Switzerland (SUI)        27    47    73    65   
123                                   Syria (SYR)        12     1     1     1   
124             Chinese Taipei (TPE) [TPE] [TPE2]        13     2     7    12   
125                              Tajikistan (TJK)         5     0     1     2   
126                          Tanzania (TAN) [TAN]        12     0     2     0   
127                                Thailand (THA)        15     7     6    11   
128                                    Togo (TOG)         9     0     0     1   
129                                   Tonga (TGA)         8     0     1     0   
130               Trinidad and Tobago (TRI) [TRI]        16     2     5    11   
131                                 Tunisia (TUN)        13     3     3     4   
132                                  Turkey (TUR)        21    39    25    24   
133                                  Uganda (UGA)        14     2     3     2   
134                                 Ukraine (UKR)         5    33    27    55   
135                    United Arab Emirates (UAE)         8     1     0     0  

### 1.3 Changer le nom des colonnes en utilisant le dictionnaire suivant :
```python
new_names =  {'Unnamed: 0': 'Country',
              '? Summer': 'Summer Olympics',
              '01 !': 'Gold',
              '02 !': 'Silver',
              '03 !': 'Bronze',
              '? Winter': 'Winter Olympics',
              '01 !.1': 'Gold.1',
              '02 !.1': 'Silver.1',
              '03 !.1': 'Bronze.1',
              '? Games': '# Games', 
              '01 !.2': 'Gold.2',
              '02 !.2': 'Silver.2',
              '03 !.2': 'Bronze.2'}


```



In [74]:
new_names =  {'Unnamed: 0': 'Country',
              '? Summer': 'Summer Olympics',
              '01 !': 'Gold',
              '02 !': 'Silver',
              '03 !': 'Bronze',
              '? Winter': 'Winter Olympics',
              '01 !.1': 'Gold.1',
              '02 !.1': 'Silver.1',
              '03 !.1': 'Bronze.1',
              '? Games': '# Games', 
              '01 !.2': 'Gold.2',
              '02 !.2': 'Silver.2',
              '03 !.2': 'Bronze.2'}
olympics_df.rename(columns=new_names, inplace=True)









olympics_df.head(5)

0         1     2     3     4      5         6     7     8  \
0                NaN  ? Summer  01 !  02 !  03 !  Total  ? Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  ? Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

In [ ]:
olympics_df.head()